Implement StreamMuxer trait for Connection

melekes opened the pull request:
- [x] StreamMuxer trait
- [x] Authenticate peer via Noise (single data channel, which can be reused later; noise can be dropped afterwards)
- [x] Write tests
- [x] https://github.com/webrtc-rs/sctp/pull/14

melekes committed:
start impl StreamMuxer trait

melekes committed:
implement poll_outbound fn

melekes committed:
implement poll fns

melekes committed:
impl close and flush_all fns

melekes committed:
wrap fields in arc and clone it in open_outbound

Fixes:

```
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to
satisfy a `'static` lifetime requirement
```

melekes committed:
implement poll_event fn

melekes committed:
add some comments

melekes committed:
close `incoming_data_channels_rx` in close

to prevent adding new data channels

melekes commented:
@tomaka could you please take a look at this when you have the time? Thanks.

tomaka commented on src/connection.rs:
```diff
@@ -20,59 +20,297 @@
+    }
+
+    fn destroy_outbound(&self, _s: Self::OutboundSubstream) {
+        // noop
```
One thing to point out is that the future of `_s` might have been polled a bit but not completely. In other words, the future might currently be waiting at one of the `await` points, and dropping the future will abruptly interrupt the execution. I don't really know whether it's a problem, though.

tomaka replied:
I would suggest adding a comment explaining this in `open_outbound`, because even if it's fine now it's something that someone who might touch the code later needs to be aware of

tomaka commented on src/connection.rs:
```diff
@@ -20,59 +20,297 @@
+            Some(detached) => {
+                let ch = PollDataChannel::new(detached);
+
+                let mut channels = self.inner.data_channels.lock().unwrap();
```
It feels a bit dangerous to me to lock a mutex here while the `rx` mutex is still locked. Can you not drop the `rx` lock before?

tomaka commented:
The multiple mutexes feel a bit dangerous to me, both in terms of deadlock and in terms of race conditions (what if another method is called between the moment a mutex is unlocked and the next one locked? is it ok everywhere?)
I think that a single mutex covering all the fields is much safer and probably not much slower.

tomaka commented:
Another point is that I would suggest keeping the single data stream implementation on the side.
I feel like [I'm getting ignored](https://github.com/libp2p/specs/pull/412#discussion_r877141696), but to me it's a good idea to also give the possibility to have this single data stream system.

melekes commented:
> I think that a single mutex covering all the fields is much safer and probably not much slower.

This is how it was done before. Unfortunately, I had to switch to multiple mutexes (futures::Mutex and std::Mutex) because I was not able to get both async and sync code to compile with just one (I've tried both futures::Mutex and std::sync::Mutex). 

If I use futures::Mutex, it works inside async future, but I had to poll it in sync code and I wasn't sure how it will play with the rest of the code (i.e. if polling future returns ready, but mutex is not ready, will the future returns ready next time it's called?).

If I use std::sync::Mutex, I have problems when locking data across await, I believe.

I can try again though.

melekes committed:
add comment to destroy_outbound

melekes committed:
drop locks early

melekes commented:
> Another point is that I would suggest keeping the single data stream implementation on the side. I feel like [I'm getting ignored](https://github.com/libp2p/specs/pull/412#discussion_r877141696), but to me it's a good idea to also give the possibility to have this single data stream system.

It's still possible to do with the new implementation. You get a `Connection` and call `open_outbound` substream. Then you use it to communicate with the other party. Or you can open additional data channels (it's up to you).

melekes committed:
reduce the number of locks

melekes committed:
bring back noise and start smoke tests

melekes committed:
reserve 1 for the initial substream

melekes committed:
smoke tests: add webrtc fingerprint to dialing addr

melekes committed:
use tokio executor in swarm

melekes committed:
do not assume fingerprint to be last

actually it's supposed to be 3rd (ip, port, fingerprint, peer_id)
but let's not assume anything

melekes committed:
increase read buf size to 8192

and reduce one in smoke tests

melekes committed:
verify peer_id and close initial stream (used for noise)

melekes committed:
allow changing read buf cap in Connection

melekes committed:
use Default::default when constructing RTCDataChannelInit

melekes committed:
remove unnecessary field from RTCDataChannelInit

melekes committed:
remove next_outbound_channel_id

It turned out we actually don't need it. webrtc-rs lib will increase
channel ID for us (on both ends on the connection).

Also, extract `register_data_channel_open_handler` fn to reduce the
amount of duplicated code.

melekes committed:
fix some clippy warnings

mxinden commented on src/transport.rs:
```diff
@@ -428,16 +410,61 @@ impl WebRTCDirectTransport {
             ))
+        };
+
+        trace!("noise handshake with {}", remote);
+        let dh_keys = Keypair::<X25519Spec>::new()
+            .into_authentic(&self.id_keys)
+            .unwrap();
+        let noise = NoiseConfig::xx(dh_keys);
+        let info = noise.protocol_info().next().unwrap();
+        // after noise is successful and we've authenticated the remote peer, encrypted IO is no
+        // longer needed, hence ignored here.
+        let (peer_id, _) = noise
+            .upgrade_outbound(PollDataChannel::new(detached), info)
+            .and_then(|(remote, io)| match remote {
+                RemoteIdentity::IdentityKey(pk) => future::ok((pk.to_peer_id(), io)),
+                _ => future::err(NoiseError::AuthenticationFailed),
+            })
+            .await
+            .map_err(Error::Noise)?;
+
+        // Verify peer's identity.
+        let peer_id_from_addr = PeerId::try_from_multiaddr(&addr);
+        if peer_id_from_addr.is_none() || peer_id_from_addr.unwrap() != peer_id {
+            error!(
+                "peer_id_from_addr {:?} != peer_id {:?}",
+                peer_id_from_addr, peer_id
+            );
+            return Err(Error::InvalidMultiaddr(addr.clone()));
         }
+
+        Ok((peer_id, Connection::new(peer_connection).await))
```
Is this sufficient to authenticate the remote and be able to fall back to the underlying WebRTC connection afterwards? Would one not need to exchange the TLS certificate fingerprint in order to prevent a monster-in-the-middle attack on the DTLS secured connection?

melekes replied:
> Is this sufficient to authenticate the remote and be able to fall back to the underlying WebRTC connection afterwards? Would one not need to exchange the TLS certificate fingerprint in order to prevent a monster-in-the-middle attack on the DTLS secured connection?

are you asking if we need to verify TLS certificate fingerprints (which we've verified during the DTLS handshake earlier) on top of Noise? Not a security expert here, but even if someone is trying to MiMA DTLS connection (which is possible), I can't see how they will be able to fake `peer_id`. So if our identity system is based on `peer_id`, then I believe we can be certain that the peer is the owner of public key associated with given `peer_id`.

mxinden replied:
Say that _A_ wants to connect to _B_. _B_ advertises the `Multiaddr` (`/ip6/<b-ip>/udp/xxx/webrtc/<b-hash>/xxx/p2p/<b-peer-id>`).

Eve intercepts B's `Multiaddr` and forwards an altered `Multiaddr` to _A_, namely `/ip6/<e-ip>/udp/xxx/webrtc/<e-hash>/xxx/p2p/<b-peer-id>` (note `e-ip` and `e-hash`).

_A_ establishes a WebRTC connection to _E_ thinking it is talking to _B_. On the established connection _A_ initiates a Noise handshake. _E_ forwards the noise handshake payloads to _B_. Once the Noise handshake succeeds _A_ falls back to the underlying WebRTC connection, thinking it is talking to _B_ instead of _E_.

By exchanging the TLS certificate fingerprints in the Noise handshake, the above attack is prevented.

Does this make sense @melekes?

melekes replied:
It does 👍 Good thinking. I will make the necessary changes

melekes committed:
exchange TLS fingerprint certificates inside noise

Fixes
https://github.com/melekes/libp2p-webrtc-direct/pull/2#discussion_r889028845

melekes committed:
remove udp_mux files and update PollDataChannel

based on new ice, sctp and data crates

melekes committed:
include old code
